users/helauth/clients.py

```diff
@@ -44,24 +49,219 @@ class HelsinkiProfileClient(BaseExternalServiceClient):
+        if user.is_anonymous:
+            return None
+
+        session_data = get_session_data(request)
+        api_tokens: dict[str, str] | None = session_data.get("api_tokens")
+        if api_tokens is None:
+            msg = "No api-tokens in session. User is not a helsinki profile user."
+            logger.info(msg)
+            return None
+
+        token: str | None = api_tokens.get(settings.OPEN_CITY_PROFILE_SCOPE)
+        if token is not None:
+            payload: ProfileTokenPayload = get_jwt_payload(token)
+            leeway = settings.HELSINKI_PROFILE_TOKEN_EXPIRATION_LEEWAY_SECONDS
+            is_token_valid = payload["exp"] > int(time.time()) + leeway
+            if is_token_valid:
+                return token
+
+        # If token or profile was not found in api-tokens, or token was expired, refresh it.
+        return cls.refresh_token(request)
+
+    @classmethod
+    def refresh_token(cls, request: WSGIRequest) -> str | None:
+        """
+        Refresh helsinki profile token.
+
+        :return: The new access token if refresh was successful, None otherwise.
+        """
+        user: AnyUser = request.user
+        if user.is_anonymous:
+            return None
+
+        access_token = TunnistamoClient.get_token(request)
+        if access_token is None:
+            msg = "Unable to get tunnistamo access token to refresh profile token."
+            logger.info(msg)
+            return None
+
+        url = f"{settings.TUNNISTAMO_BASE_URL}/api-tokens/"
+        headers: dict[str, str] = {"Authorization": f"Bearer {access_token}"}
+
+        response = cls.post(url=url, headers=headers)
+
+        if response.status_code != 200:
+            msg = (
+                f"Unable to get API tokens for helsinki profile for user {int(request.user.pk)}: "
+                f"[{response.status_code}] {response.text}"
+            )
+            logger.info(msg)
+            return None
+
+        session_data = get_session_data(request)
+        session_data["api_tokens"] = BaseExternalServiceClient.response_json(response=response)
+        return session_data["api_tokens"].get(settings.OPEN_CITY_PROFILE_SCOPE)
+
+    @classmethod
+    def graphql_request(cls, request: WSGIRequest, *, query: str) -> MyProfileData | None:
         token = cls.get_token(request)
         if token is None:
             return None
 
         url: str = settings.OPEN_CITY_PROFILE_GRAPHQL_API
-        request_data = {"query": cls._reservation_prefill_query()}
+        request_data = {"query": query}
         headers = {"Authorization": f"Bearer {token}"}
 
         response = cls.post(url=url, json=request_data, headers=headers)
         response_data = cls.response_json(response)
 
-        my_profile_data: MyProfileData = get_nested(response_data, "data", "myProfile", default={})
-        return ReservationPrefillParser(my_profile_data).parse()
+        return get_nested(response_data, "data", "myProfile", default={})
```

Comment:
Is it `myProfile` in the response if other user's profile is queried?

Reply:
Yes, this should be configurable.